inc/zoo/swar/SWAR.h

```diff
@@ -430,13 +430,49 @@ booleans(SWAR<NB, T> arg) noexcept {
 template<int NBits, typename T>
 constexpr auto
 differents(SWAR<NBits, T> a1, SWAR<NBits, T> a2) {
-    return booleans(a1 ^ a2);
+    return ~equals(a1, a2);
 }
 
+/*
+ * Returns a BooleanSWAR that contains a true value in each lane where the two
+ * inputs SWARs match.
+ */ 
 template<int NBits, typename T>
 constexpr auto
 equals(SWAR<NBits, T> a1, SWAR<NBits, T> a2) {
-    return ~differents(a1, a2);
+    using S = swar::SWAR<NBits, T>;
+    const auto h = S::MostSignificantBit, l = S::LeastSignificantBit;
+    const auto x = (a1 ^ a2).value();
+    return BooleanSWAR<NBits, T>(h & ~(x| ((x|h) -1)));
```

Comment:
This code seems to have a mistake: subtracting the number 1 instead of `l` (small caps L).  You would not have had that problem if you would have used SWAR instead of unsigned operators (`x` is an unsigned, not a SWAR).

I don't doubt that this is correct (substituting 1 by `l`), but this code does not attempt to explain what is happening.
See the other implementations of SWAR operations:
1. There is an effort to explain the intermediate results
2. In the more modern implementations, the effort intensified, for example by giving a name to each partial result or each stage of the technique.

Regardless that probably `x` is what TAOCP names this, we can say `auto differentLanes = a1 ^ a2;`
In the same way we do it all over the codebase, we don't say `h` but MSB or something to that effect.

Reply:
Subtracting 1 is an error, I think.
How utterly strange that I ran the unit tests on my machine on this PR and they passed, giving me confidence there wasn't an error.  Could it be that our unit tests around equals() and differents() are somehow lacking? How strange.  Seems like there should be really strong tests around such a basic operation.

The intermediate results here are fairly uncomprehensible.  We're doing something with the same flavor as the xor swap temporaries technique, and most people don't have good intuition for 'ok let's smear these test values with the high/ lo boundaries of their ranges'.  We could go for a full explanation towards greater equal (hey, this is actually a strange edge case of less than zero that requires less instructions, combined with the idea that we can easily make nonzero bytes with a xor of the original test values).

I'm not even sure that's a good way of describing it!  Would some theorem prover given a more general phrasing of this operation even come up with these operands for all hardware it'll execute on? I don't know!  It seems ok to present it as working without having a good set of names for the intermediates of the operations involved.

Reply:
I did not write good unit tests when I was the sole contributor.  This is because of my laziness, stupidity, and because of the certainty that I gotten correct implementations obtained through other means.  You're now the victim of my unit test mistakes.
Still, please notice that if you would have used the SWAR `operator-` instead of the integer one, you'd have had a compilation error.
If you find, during normal development, that the SWAR operations get in the way of getting things done, please let me know, those are failures of ergonomics to be corrected.  If you're in a hurry, do what you have to do.
There is no doubt subtracting the *number* 1 is an error, since it affects only the least significant lane, because of the work on all the MSBs.
What do you suggest we do with preexisting unit tests?

Reply:
Ah, we can't explain what we can't understand.  Proving that the technique works is beside the point.  Just at least indicate that the terse names are because those are the names used in TAOCP

Reply:
I'll extend testing before submitting this PR now that I've had a very direct indication that there is a huge hole.